src/main/scala/com/mongodb/spark/sql/v2/MongoDataSourceWriter.scala

```diff
@@ -0,0 +1,121 @@
+import com.mongodb.spark.sql.MapFunctions.rowToDocumentMapper
+import com.mongodb.spark.sql.v2.MongoDataSourceWriter.SaveType
+import org.apache.spark.sql.sources.v2.writer.{DataSourceWriter, DataWriter, DataWriterFactory, WriterCommitMessage}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{Row, SaveMode}
+import org.bson.BsonDocument
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+object MongoDataSourceWriter {
+  object SaveType extends Enumeration {
+    type SaveType = Value
+    val Insert, Update, Replace = Value
+  }
+}
+
+case class MongoDataSourceWriter(private val writeConfig: WriteConfig, private val schema: StructType, private val mode: SaveMode) extends DataSourceWriter {
+  private val mapper = rowToDocumentMapper(schema)
+  private val fieldNames = schema.fieldNames.toList
+  private val queryKeyList = BsonDocument.parse(writeConfig.shardKey.getOrElse("{_id: 1}")).keySet().asScala.toList
+  private val containsQueryKeyList = queryKeyList.forall(fieldNames.contains)
+  private val saveType = {
+    if (mode == SaveMode.Overwrite || writeConfig.forceInsert) {
+      SaveType.Insert
+    } else if (containsQueryKeyList) {
+      if (writeConfig.replaceDocument) SaveType.Replace else SaveType.Update
+    } else {
+      SaveType.Insert
+    }
+  }
+  val mongoConnector = MongoConnector(writeConfig.asOptions)
+
+  override def createWriterFactory(): DataWriterFactory[Row] = MongoDataWriteFactory()
+
+  override def commit(messages: Array[WriterCommitMessage]): Unit = {
+  }
+
+  override def abort(messages: Array[WriterCommitMessage]): Unit = {
+  }
+
+  private case class MongoDataWriteFactory() extends DataWriterFactory[Row] {
+    override def createDataWriter(partitionId: Int, attemptNumber: Int): DataWriter[Row] = MongoDataWriter(partitionId, attemptNumber)
+  }
+
+  private case class MongoDataWriter(partitionId: Int, attemptNumber: Int) extends DataWriter[Row] {
+    @transient private val batch: mutable.Queue[WriteModel[BsonDocument]] = mutable.Queue.empty[WriteModel[BsonDocument]]
+    @transient private var mongoClient: Option[MongoClient] = None
+
+    override def write(record: Row): Unit = {
+      val doc = mapper(record)
+      val op = if (saveType == SaveType.Insert || !queryKeyList.forall(doc.containsKey)) {
+        new InsertOneModel[BsonDocument](doc)
+      } else {
+        val queryDocument = new BsonDocument()
+        queryKeyList.foreach(key => queryDocument.append(key, doc.get(key)))
+        if (saveType == SaveType.Replace) {
+          new ReplaceOneModel[BsonDocument](queryDocument, doc, new ReplaceOptions().upsert(true))
+        } else {
+          queryDocument.keySet().asScala.foreach(doc.remove(_))
+          new UpdateOneModel[BsonDocument](queryDocument, new BsonDocument("$set", doc), new UpdateOptions().upsert(true))
+        }
+      }
+      writeOperation(op)
+    }
+
+    override def commit(): WriterCommitMessage = {
```

Comment:
So MongoDB can't do transactions with > 16MB even on 4.0, but it seems like we're violating the visibility guarantees documented in the Javadoc for API being implemented here.

Reply:
OK - I see its not transactional in that nature - eg the whole operation is a single transaction. So we don't need ACID guarantees of using Mongo Transactions. But it does need a refactor.